website/server/libs/challenges/handleGetMembersForChallenge.js

```diff
@@ -0,0 +1,92 @@
+import _ from 'lodash';
+import {
+  model as User,
+  publicFields as memberFields,
+  nameFields,
+} from '../../models/user';
+import { model as Challenge } from '../../models/challenge';
+import { model as Group } from '../../models/group';
+import * as Tasks from '../../models/task';
+import { NotFound } from '../errors';
+
+async function getMembersTasksForChallenge (members, challenge) {
+  const challengeTasks = await Tasks.Task.find({
+    userId: { $in: members.map(m => m.id) },
+    'challenge.id': challenge.id,
+  })
+    .select('-tags -checklist') // We don't want to return tags and checklists publicly
+    .lean()
+    .exec();
+
+  return _.groupBy(challengeTasks, 'userId');
+}
+
+export async function handleGetMembersForChallenge (req, res) {
+  req.checkParams('challengeId', res.t('challengeIdRequired')).notEmpty().isUUID();
+  req.checkQuery('lastId').optional().notEmpty().isUUID();
+  // Allow an arbitrary number of results (up to 60)
+  req.checkQuery('limit', res.t('groupIdRequired')).optional().notEmpty().isInt({ min: 1, max: 60 });
+
+  const validationErrors = req.validationErrors();
+  if (validationErrors) throw validationErrors;
+
+  const { challengeId } = req.params;
+  const { lastId } = req.query;
+  const { user } = res.locals;
+
+  const challenge = await Challenge.findById(challengeId).select('_id type leader group').exec();
+  if (!challenge) throw new NotFound(res.t('challengeNotFound'));
+
+  // optionalMembership is set to true because even
+  // if you're not member of the group you may be able to access the challenge
+  // for example if you've been booted from it, are the leader or a site admin
+  const group = await Group.getGroup({
+    user,
+    groupId: challenge.group,
+    fields: '_id type privacy',
+    optionalMembership: true,
+  });
+
+  if (!group || !challenge.canView(user, group)) throw new NotFound(res.t('challengeNotFound'));
+
+  const query = {};
+  let fields = nameFields;
+  // add computes stats to the member info when items and stats are available
+  let addComputedStats = false;
+  query.challenges = challenge._id;
+
+  if (req.query.includeAllPublicFields === 'true') {
+    fields = memberFields;
+    addComputedStats = true;
+  }
+
+  if (req.query.search) {
+    query['auth.local.username'] = { $regex: req.query.search };
+  }
+
+  if (lastId) query._id = { $gt: lastId };
+
+  let limit = req.query.limit ? Number(req.query.limit) : 30;
+
+  // Allow for all challenges members to be returned
+  if (req.query.includeAllMembers === 'true') {
+    limit = 0; // no limit
+  }
+
+  const members = await User
+    .find(query)
+    .sort({ _id: 1 })
+    .limit(limit)
+    .select(fields)
+    .lean()
+    .exec();
+
+  members.forEach(member => User.transformJSONUser(member, addComputedStats));
+  if (req.query.includeTasks === 'true') {
+    const memberIdToTasksMap = await getMembersTasksForChallenge(members, challenge);
+    for (const member of members) {
```

Comment:
instead of having two loops here could you add the tasks to the member inside the first forEach?

```javascript
  const includeTasks = req.query.includeTasks === 'true';
  let memberIdToTasksMap;
  if (includeTasks) {
    memberIdToTasksMap = await getMembersTasksForChallenge(members, challenge);
  }
  members.forEach(member => {
       User.transformJSONUser(member, addComputedStats);
       if (includeTasks) member.tasks = memberIdToTasksMap[member._id]; // id might not be returned when using lean()
  });
```

something like this

Reply:
It's up to you to decide, but maybe it would be better to have the `includeTasks`-related logic to be enclosed in a separate block? The algorithmic complexity is still the same with both solutions, I just thought that it might be a tiny bit more readable this way.

Reply:
The big-O complexity is indeed the same but since we're talking about small numbers having 2 separate loops can result in an important performance hit so i'd prefer if we could limit the loops to one, thanks!

Reply:
Ok, no problem :)